File: JSTerm/JSKern/Communication/DBus.m
//
//  DBus.m
//  JSTerm
//
//  Created by fridakitten on 14.03.25.
//

#import "DBus.h"

/*
 @Brief for each DBUS connection a process has
 */
@implementation JS_DBUS

- (instancetype)init
{
    self = [super init];
    _semaphore = NULL;
    _data = @"";
    return self;
}

- (void)attach:(dispatch_semaphore_t)semaphore
{
    _semaphore = semaphore;
}

- (NSString*)waitformsg
{
    if(_semaphore == NULL)
    {
        return @"";
    }
    
    dispatch_semaphore_wait(_semaphore, DISPATCH_TIME_FOREVER);
    
    return _data;
}

- (void)sendmsg:(NSString*)msg
{
    if(_semaphore == NULL)
    {
        return;
    }
    
    _data = msg;
    
    dispatch_semaphore_signal(_semaphore);
}

@end

/*
 @Brief the system that manages the DBus connections
 */
@implementation JS_DBUS_SYSTEM

- (instancetype)init
{
    self = [super init];
    _bus = [NSMutableDictionary dictionary];
    return self;
}

- (void)register:(NSString*)bus_id
{
    _bus[bus_id] = [[JS_DBUS alloc] init];
}

- (void)unregister:(NSString*)bus_id
{
    [_bus removeObjectForKey:bus_id];
}

- (NSString*)waitformsg:(NSString*)bus_id semaphore:(dispatch_semaphore_t) semaphore
{
    [_bus[bus_id] attach:semaphore];
    return [_bus[bus_id] waitformsg];
}

- (void)sendmsg:(NSString*)bus_id payload:(NSString*)payload
{
    [_bus[bus_id] sendmsg:payload];
}

@end

File: JSTerm/JSKern/Communication/Dbus.swift
/*
Dbus.swift
 
Copyright (C) 2024 fridakitten

This file is part of JSTerm.

FridaCodeManager is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

FridaCodeManager is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with FridaCodeManager. If not, see <https://www.gnu.org/licenses/>.

 ______    _     _         _____        __ _                           ______                    _       _   _
|  ___|  (_)   | |       /  ___|      / _| |                          |  ___|                  | |     | | (_)
| |_ _ __ _  __| | __ _  \ `--.  ___ | |_| |___      ____ _ _ __ ___  | |_ ___  _   _ _ __   __| | __ _| |_ _  ___  _ __
|  _| '__| |/ _` |/ _` |  `--. \/ _ \|  _| __\ \ /\ / / _` | '__/ _ \ |  _/ _ \| | | | '_ \ / _` |/ _` | __| |/ _ \| '_ \
| | | |  | | (_| | (_| | /\__/ / (_) | | | |_ \ V  V / (_| | | |  __/ | || (_) | |_| | | | | (_| | (_| | |_| | (_) | | | |
\_| |_|  |_|\__,_|\__,_| \____/ \___/|_|  \__| \_/\_/ \__,_|_|  \___| \_| \___/ \__,_|_| |_|\__,_|\__,_|\__|_|\___/|_| |_|
Founded by. Sean Boleslawski, Benjamin Hornbeck and Lucienne Salim in 2023
*/

import Foundation

class JS_DBUS {
    private var semaphore: DispatchSemaphore?
    private var data: String = ""
    
    func attachsemaphore(semaphore: DispatchSemaphore) -> Void {
        self.semaphore = semaphore
    }
    
    func waitformsg() -> String {
        semaphore?.wait()
        return data
    }
    
    func sendmsg(payload: String) {
        data = payload
        semaphore?.signal()
    }
}

class JS_DBUS_SYSTEM {
    private var bus: [String:JS_DBUS] = [:]
    
    func register(id: String) {
        bus[id] = JS_DBUS()
    }
    
    func unregister(id: String) {
        bus.removeValue(forKey: id)
    }
    
    func waitformsg(semaphore: DispatchSemaphore, id: String) -> String {
        guard let idbus: JS_DBUS = bus[id] else { return "" }
        idbus.attachsemaphore(semaphore: semaphore)
        return idbus.waitformsg()
    }
    
    func sendmsg(id: String, payload: String) {
        guard let idbus: JS_DBUS = bus[id] else { return }
        idbus.sendmsg(payload: payload)
    }
}
